### ISSUE TITLE
[BUG] secureJsonData are not preserved on upload

### ISSUE THREAD

**Issue by kossioni** [esnet/gdg#229]
**Describe the bug**
You download a connection and of course it doesn't contain username/passwd and of course there is a configuration to supply credentials in connection basic auth.  But when you use influxdb connection a token is required on top of basic auth. So when we change the json downloaded and put secureJsonData: {token: "123"} this is not preserved when we upload the connection to another grafana.

<img width="743" alt="image" src="https://github.com/esnet/gdg/assets/101812286/91450eb6-3978-4a8b-9b6d-a64b04ccf2ae">

You need to preserve the secureJsonData because we can put additional stuff in there.

**Comment by csg33k** [comment]
Okay, that's fair.  Hmm.  


Trying to think of a clean pattern for this.  

I like the importer pattern for auth, since it allows you to manage say 20+ connections and easily update the auth in one place.  Though I don't like the idea of having secureData in two locations I'm not a huge fan of.  It'll just end up with fighting out to see who is 'right' / source of truth.

How does this sound.

1.  If secureJsonData exists, use it instead of importer.yml data.  That includes auth information.
2. Allow importer.yml to set other properties ie. dynamic secureJson via rule matching similar to auth, but support an unstructured map, that it will feed as secure data.

I think that should support most use cases.  SecureData is stripped away on download, so the only way the JSON should have data is if it was manually added in by the user.

**Comment by csg33k** [comment]
This is a bit a brain dump, so.. I apologize if it's a bit non-sensical.  

---
This is a bit more convoluted that I had expected.  I have a branch that mostly works, but due to how viper handles config loading, it won't work as expected.  

viper forces a lowercase on all the keys it's reads from the configuration.  Which generates broken JSON since keys like. 

```json
"httpHeaderValue1": "token Value"
```

turn into

```json
"httpheadervalue1": "token Value"
```

which grafana disregards since it's an invalid json field we're setting.  

Additionally and this is a bit of a lesser issue, but the header field is part of the JSON, so you may end up with a bit of a mismatch over time which isn't ideal.  

here's an example JSON, truncated for readability

```
{
  "access": "proxy",
  "basicAuth": true,
  "basicAuthUser": "user",
  "database": "om-netsage-irnc-*",
  "id": 12,
  "isDefault": true,
  "jsonData": {
    "esVersion": 70,
    "httpHeaderName1": "Moo",
    "httpHeaderName2": "Foo",
  },
  "name": "netsage",
  "orgId": 1,
  "readOnly": false,
  "secureJsonData": {
    "httpHeaderValue2": "Bar"
  },
  "secureJsonFields": {
        //managed by importer.yml
  },
  "type": "elasticsearch",
  "typeLogoUrl": "",
  "uid": "rg9qPqP7z",
  "url": "https://netsage-elk1.grnoc.iu.edu/esproxy2/",
  "user": "",
  "version": 2,
  "withCredentials": false
}
```

initially, we can assume that the value we set in the importer.yml matches Moo, but over time if someone removed Moo, and we're importing it again, it may have unexpected results.  I'm not sure if it'll set the value of Foo to Moo, or what the behavior is when it passes a headerName1=Foo and headerValue1=ValueOfMoo.  

I can do a bandaid solution for now, an just add 'token' as a supported field but it's going to basically be a short list of supported secure fields.  (Basic Auth + Token)

Alternatively, which is probably where i'm heading towards right now.

is have a predefined path like "secure" or "credentials" which stores all of your authentication as a JSON/Yaml file.  

Then the rules that previous set basic auth would instead map to a file location which is loaded by GDG and used.  It's not ideal, and naming convention is really left to the user, can be reused, etc.  

This would get a bit complicated with cloud support as GDG assumes all assets are on S3/GCS when that remote storage is turned on, but we can probably over look that for that specific use case.

Any thoughts.  @kossioni on a preferred approach for your use case at least.. no promises on what approach will be taken but feed back is always appreciated.

**Comment by kossioni** [comment]
I see that there is quite some trouble in the viper thread here with case sensitivity: https://github.com/spf13/viper/issues/260

What about this:

connections:
          credential_rules:
            - rules:
                - field: "name"
                  regex: "^Prometheus$"
              auth:
                user: admin
                password: Admin4Prometheus
            - rules:
                - field: "name"
                  regex: "^PrometheusCustom$"
              auth:
                custom: "ewogICJzZWN1cmVKc29uRGF0YSI6IHsKICAgICJ0bHNDQUNlcnQiOiAiLi4uIiwKICAgICJ0bHNDbGllbnRDZXJ0IjogIi4uLiIsCiAgICAidGxzQ2xpZW50S2V5IjogIi4uLiIsCiAgICAicGFzc3dvcmQiOiBudWxsLAogICAgImJhc2ljQXV0aFBhc3N3b3JkIjogIkFkbWluNEluZmx1eFRlc3QiLAogICAgInRva2VuIjogIlZZczNnSDJBWFFHOHc0Z0VOTG1NTkJWTkcyVlNNTUlUU09TaHRBYVFGbW0tVFBkOVF4dElCSlpmNWQ5RkJwN1NVUDdWWTFUanJPajRlbmc9PSIKICB9Cn0="

So when auth has the custom key you will read the base64 value and decode in memory and you will get the secureJsonData:

{
  "secureJsonData": {
    "tlsCACert": "...",
    "tlsClientCert": "...",
    "tlsClientKey": "...",
    "password": null,
    "basicAuthPassword": "Admin4InfluxTest",
    "token": "VYs3gH2AXQG8w4gENLmMNBVNG2VSMMITSOShtAaQFmm-TPd9QxtIBJZf5d9FBp7SUP7VY1TjrOj4eng=="
  }
}

You will pick the secureJsonData and put as is to the json before datasource is uploaded to grafana.

I think that will work either in k8s or in standalone installations and it is somewhat more secure that having the credentials  dangling in free text as well.

I don't think viper will now mess with the case sensivity of a base64 value and everybody will be happy. :)

How does it sound to you?

**Comment by safaci2000** [comment]
This is where I wish GDG had some pluginin system you could leverage.  I keep revisiting options for pre-post processing, but even for auth all these patterns seem okay but not sure I'd want to have it as part of the standard flow.  I'm also not a fan of adding too many supported patterns since usually we end with some code path not fully tested as the others.

I think that maintains the overall credential_rule logic is it works today, which is good.  I get the base64 gets past the limitations of viper, which is good but I don't see that as improving security or usability. 

if you don't live in K8 24/7 having JSON blob encoded as base64 is a pretty peculiar pattern from a usability perspective.  Also, if you end up using a file or a base64 blob, wouldn't you still load the importer.yml, secureData etc all in a K8 secret which can be handled by hashicorp or whatever secret manager you use?  Or provide ENV values that override the corresponding settings from the importer.yml?

I think expecting the user to take the JSON in base64 makes this a good bit more complicated.  It already feels like GDG has too many patterns that need a good bit more clarification in the docs, I'm a bit reluctant to add yet one more pattern like that. 

Additionally, the bits you're encoding in the base64 are likely more than I'd like to support.  basicAuth, withCredentials should come from grafana when it pulls the DS, having encoded data that overrides those values will cause ambiguity and make it hard to debug.  I'd say the only thing that the custom should contain is the secureJsonData.

**Comment by safaci2000** [comment]
@kossioni this change has been released v0.5.2 went out today.
